rspace/src/main/scala/coop/rchain/rspace/history/HistoryRepositoryImpl.scala

```diff
@@ -183,22 +184,24 @@ final case class HistoryRepositoryImpl[F[_]: Sync, C, P, A, K](
         (key, None, DeleteAction(key.bytes.toSeq.toList))
     }
 
-  private def storeLeaves(leafs: List[Result]): F[List[HistoryAction]] =
-    leafs.traverse {
-      case (key, Some(data), historyAction) =>
-        leafStore.put(key, data).map(_ => historyAction)
-      case (_, None, historyAction) =>
-        Applicative[F].pure(historyAction)
-    }
+  private def storeLeaves(leafs: List[Result]): F[List[HistoryAction]] = {
+    val toBeStored = leafs.filter(_._2.isDefined).map(v => (v._1, v._2.get))
+    leafStore.put(toBeStored).map(_ => leafs.map(_._3))
+  }
 
-  override def checkpoint(actions: List[HotStoreAction]): F[HistoryRepository[F, C, P, A, K]] =
+  override def checkpoint(actions: List[HotStoreAction]): F[HistoryRepository[F, C, P, A, K]] = {
+    implicit val P = Par[F].parallel
```

Comment:
This could be moved to class level

Reply:
Im torn here. on one hand I agree. on the other I like the locality of Parallel.